File: packages/MeeshySDK/Sources/MeeshySDK/Models/LanguageData.swift
import Foundation

public struct LanguageInfo: Sendable {
    public let code: String
    public let name: String
    public let nativeName: String
    public let flag: String
    public let colorHex: String

    public init(code: String, name: String, nativeName: String, flag: String, colorHex: String) {
        self.code = code
        self.name = name
        self.nativeName = nativeName
        self.flag = flag
        self.colorHex = colorHex
    }
}

public enum LanguageData {
    // MARK: - All Languages (translation targets)

    public static let allLanguages: [LanguageInfo] = [
        // Romance
        LanguageInfo(code: "fr", name: "French", nativeName: "Français", flag: "\u{1F1EB}\u{1F1F7}", colorHex: "3B82F6"),
        LanguageInfo(code: "es", name: "Spanish", nativeName: "Español", flag: "\u{1F1EA}\u{1F1F8}", colorHex: "EF4444"),
        LanguageInfo(code: "it", name: "Italian", nativeName: "Italiano", flag: "\u{1F1EE}\u{1F1F9}", colorHex: "22C55E"),
        LanguageInfo(code: "pt", name: "Portuguese", nativeName: "Português", flag: "\u{1F1E7}\u{1F1F7}", colorHex: "16A34A"),
        LanguageInfo(code: "ro", name: "Romanian", nativeName: "Română", flag: "\u{1F1F7}\u{1F1F4}", colorHex: "2563EB"),

        // Germanic
        LanguageInfo(code: "en", name: "English", nativeName: "English", flag: "\u{1F1EC}\u{1F1E7}", colorHex: "6366F1"),
        LanguageInfo(code: "de", name: "German", nativeName: "Deutsch", flag: "\u{1F1E9}\u{1F1EA}", colorHex: "F59E0B"),
        LanguageInfo(code: "nl", name: "Dutch", nativeName: "Nederlands", flag: "\u{1F1F3}\u{1F1F1}", colorHex: "F97316"),
        LanguageInfo(code: "sv", name: "Swedish", nativeName: "Svenska", flag: "\u{1F1F8}\u{1F1EA}", colorHex: "0EA5E9"),
        LanguageInfo(code: "da", name: "Danish", nativeName: "Dansk", flag: "\u{1F1E9}\u{1F1F0}", colorHex: "DC2626"),
        LanguageInfo(code: "no", name: "Norwegian", nativeName: "Norsk", flag: "\u{1F1F3}\u{1F1F4}", colorHex: "1D4ED8"),
        LanguageInfo(code: "af", name: "Afrikaans", nativeName: "Afrikaans", flag: "\u{1F1FF}\u{1F1E6}", colorHex: "059669"),

        // Slavic
        LanguageInfo(code: "ru", name: "Russian", nativeName: "\u{0420}\u{0443}\u{0441}\u{0441}\u{043A}\u{0438}\u{0439}", flag: "\u{1F1F7}\u{1F1FA}", colorHex: "DC2626"),
        LanguageInfo(code: "uk", name: "Ukrainian", nativeName: "\u{0423}\u{043A}\u{0440}\u{0430}\u{0457}\u{043D}\u{0441}\u{044C}\u{043A}\u{0430}", flag: "\u{1F1FA}\u{1F1E6}", colorHex: "FBBF24"),
        LanguageInfo(code: "pl", name: "Polish", nativeName: "Polski", flag: "\u{1F1F5}\u{1F1F1}", colorHex: "E11D48"),
        LanguageInfo(code: "cs", name: "Czech", nativeName: "Čeština", flag: "\u{1F1E8}\u{1F1FF}", colorHex: "1E40AF"),
        LanguageInfo(code: "sk", name: "Slovak", nativeName: "Slovenčina", flag: "\u{1F1F8}\u{1F1F0}", colorHex: "1E3A8A"),
        LanguageInfo(code: "bg", name: "Bulgarian", nativeName: "\u{0411}\u{044A}\u{043B}\u{0433}\u{0430}\u{0440}\u{0441}\u{043A}\u{0438}", flag: "\u{1F1E7}\u{1F1EC}", colorHex: "15803D"),
        LanguageInfo(code: "hr", name: "Croatian", nativeName: "Hrvatski", flag: "\u{1F1ED}\u{1F1F7}", colorHex: "B91C1C"),
        LanguageInfo(code: "sr", name: "Serbian", nativeName: "\u{0421}\u{0440}\u{043F}\u{0441}\u{043A}\u{0438}", flag: "\u{1F1F7}\u{1F1F8}", colorHex: "9F1239"),
        LanguageInfo(code: "sl", name: "Slovenian", nativeName: "Slovenščina", flag: "\u{1F1F8}\u{1F1EE}", colorHex: "0369A1"),

        // Baltic
        LanguageInfo(code: "lt", name: "Lithuanian", nativeName: "Lietuvių", flag: "\u{1F1F1}\u{1F1F9}", colorHex: "CA8A04"),
        LanguageInfo(code: "lv", name: "Latvian", nativeName: "Latviešu", flag: "\u{1F1F1}\u{1F1FB}", colorHex: "7C2D12"),

        // Finno-Ugric
        LanguageInfo(code: "fi", name: "Finnish", nativeName: "Suomi", flag: "\u{1F1EB}\u{1F1EE}", colorHex: "2563EB"),
        LanguageInfo(code: "hu", name: "Hungarian", nativeName: "Magyar", flag: "\u{1F1ED}\u{1F1FA}", colorHex: "B45309"),
        LanguageInfo(code: "et", name: "Estonian", nativeName: "Eesti", flag: "\u{1F1EA}\u{1F1EA}", colorHex: "0284C7"),

        // Hellenic
        LanguageInfo(code: "el", name: "Greek", nativeName: "\u{0395}\u{03BB}\u{03BB}\u{03B7}\u{03BD}\u{03B9}\u{03BA}\u{03AC}", flag: "\u{1F1EC}\u{1F1F7}", colorHex: "1D4ED8"),

        // Turkic
        LanguageInfo(code: "tr", name: "Turkish", nativeName: "Türkçe", flag: "\u{1F1F9}\u{1F1F7}", colorHex: "EF4444"),
        LanguageInfo(code: "az", name: "Azerbaijani", nativeName: "Azərbaycan", flag: "\u{1F1E6}\u{1F1FF}", colorHex: "0891B2"),
        LanguageInfo(code: "kk", name: "Kazakh", nativeName: "\u{049A}\u{0430}\u{0437}\u{0430}\u{049B}\u{0448}\u{0430}", flag: "\u{1F1F0}\u{1F1FF}", colorHex: "0EA5E9"),
        LanguageInfo(code: "uz", name: "Uzbek", nativeName: "O'zbek", flag: "\u{1F1FA}\u{1F1FF}", colorHex: "0D9488"),

        // Semitic
        LanguageInfo(code: "ar", name: "Arabic", nativeName: "\u{0627}\u{0644}\u{0639}\u{0631}\u{0628}\u{064A}\u{0629}", flag: "\u{1F1F8}\u{1F1E6}", colorHex: "15803D"),
        LanguageInfo(code: "he", name: "Hebrew", nativeName: "\u{05E2}\u{05D1}\u{05E8}\u{05D9}\u{05EA}", flag: "\u{1F1EE}\u{1F1F1}", colorHex: "1D4ED8"),
        LanguageInfo(code: "am", name: "Amharic", nativeName: "\u{12A0}\u{121B}\u{122D}\u{129B}", flag: "\u{1F1EA}\u{1F1F9}", colorHex: "16A34A"),

        // Indo-Aryan
        LanguageInfo(code: "hi", name: "Hindi", nativeName: "\u{0939}\u{093F}\u{0928}\u{094D}\u{0926}\u{0940}", flag: "\u{1F1EE}\u{1F1F3}", colorHex: "F97316"),
        LanguageInfo(code: "bn", name: "Bengali", nativeName: "\u{09AC}\u{09BE}\u{0982}\u{09B2}\u{09BE}", flag: "\u{1F1E7}\u{1F1E9}", colorHex: "059669"),
        LanguageInfo(code: "ur", name: "Urdu", nativeName: "\u{0627}\u{0631}\u{062F}\u{0648}", flag: "\u{1F1F5}\u{1F1F0}", colorHex: "16A34A"),
        LanguageInfo(code: "ne", name: "Nepali", nativeName: "\u{0928}\u{0947}\u{092A}\u{093E}\u{0932}\u{0940}", flag: "\u{1F1F3}\u{1F1F5}", colorHex: "DC2626"),
        LanguageInfo(code: "ta", name: "Tamil", nativeName: "\u{0BA4}\u{0BAE}\u{0BBF}\u{0BB4}\u{0BCD}", flag: "\u{1F1EE}\u{1F1F3}", colorHex: "D97706"),

        // Iranian
        LanguageInfo(code: "fa", name: "Persian", nativeName: "\u{0641}\u{0627}\u{0631}\u{0633}\u{06CC}", flag: "\u{1F1EE}\u{1F1F7}", colorHex: "059669"),

        // Caucasian
        LanguageInfo(code: "ka", name: "Georgian", nativeName: "\u{10E5}\u{10D0}\u{10E0}\u{10D7}\u{10E3}\u{10DA}\u{10D8}", flag: "\u{1F1EC}\u{1F1EA}", colorHex: "B91C1C"),
        LanguageInfo(code: "hy", name: "Armenian", nativeName: "\u{0540}\u{0561}\u{0575}\u{0565}\u{0580}\u{0565}\u{0576}", flag: "\u{1F1E6}\u{1F1F2}", colorHex: "EA580C"),

        // East Asian
        LanguageInfo(code: "zh", name: "Chinese", nativeName: "\u{4E2D}\u{6587}", flag: "\u{1F1E8}\u{1F1F3}", colorHex: "DC2626"),
        LanguageInfo(code: "ja", name: "Japanese", nativeName: "\u{65E5}\u{672C}\u{8A9E}", flag: "\u{1F1EF}\u{1F1F5}", colorHex: "E11D48"),
        LanguageInfo(code: "ko", name: "Korean", nativeName: "\u{D55C}\u{AD6D}\u{C5B4}", flag: "\u{1F1F0}\u{1F1F7}", colorHex: "1E40AF"),

        // Southeast Asian
        LanguageInfo(code: "th", name: "Thai", nativeName: "\u{0E44}\u{0E17}\u{0E22}", flag: "\u{1F1F9}\u{1F1ED}", colorHex: "7C3AED"),
        LanguageInfo(code: "vi", name: "Vietnamese", nativeName: "Tiếng Việt", flag: "\u{1F1FB}\u{1F1F3}", colorHex: "DC2626"),
        LanguageInfo(code: "id", name: "Indonesian", nativeName: "Bahasa Indonesia", flag: "\u{1F1EE}\u{1F1E9}", colorHex: "EF4444"),
        LanguageInfo(code: "ms", name: "Malay", nativeName: "Bahasa Melayu", flag: "\u{1F1F2}\u{1F1FE}", colorHex: "1D4ED8"),
        LanguageInfo(code: "tl", name: "Filipino", nativeName: "Filipino", flag: "\u{1F1F5}\u{1F1ED}", colorHex: "2563EB"),
        LanguageInfo(code: "my", name: "Burmese", nativeName: "\u{1019}\u{103C}\u{1014}\u{103A}\u{1019}\u{102C}\u{1018}\u{102C}\u{101E}\u{102C}", flag: "\u{1F1F2}\u{1F1F2}", colorHex: "CA8A04"),
        LanguageInfo(code: "km", name: "Khmer", nativeName: "\u{1797}\u{17B6}\u{179F}\u{17B6}\u{1781}\u{17D2}\u{1798}\u{17C2}\u{179A}", flag: "\u{1F1F0}\u{1F1ED}", colorHex: "1E3A8A"),
        LanguageInfo(code: "lo", name: "Lao", nativeName: "\u{0EA5}\u{0EB2}\u{0EA7}", flag: "\u{1F1F1}\u{1F1E6}", colorHex: "B91C1C"),

        // West African
        LanguageInfo(code: "yo", name: "Yoruba", nativeName: "Yorùbá", flag: "\u{1F1F3}\u{1F1EC}", colorHex: "16A34A"),
        LanguageInfo(code: "ig", name: "Igbo", nativeName: "Igbo", flag: "\u{1F1F3}\u{1F1EC}", colorHex: "15803D"),
        LanguageInfo(code: "ha", name: "Hausa", nativeName: "Hausa", flag: "\u{1F1F3}\u{1F1EC}", colorHex: "059669"),
        LanguageInfo(code: "wo", name: "Wolof", nativeName: "Wolof", flag: "\u{1F1F8}\u{1F1F3}", colorHex: "0D9488"),
        LanguageInfo(code: "bm", name: "Bambara", nativeName: "Bamanankan", flag: "\u{1F1F2}\u{1F1F1}", colorHex: "CA8A04"),
        LanguageInfo(code: "ff", name: "Fulah", nativeName: "Fulfulde", flag: "\u{1F1F8}\u{1F1F3}", colorHex: "0891B2"),
        LanguageInfo(code: "tw", name: "Twi", nativeName: "Twi", flag: "\u{1F1EC}\u{1F1ED}", colorHex: "D97706"),
        LanguageInfo(code: "ee", name: "Ewe", nativeName: "E\u{028B}egbe", flag: "\u{1F1EC}\u{1F1ED}", colorHex: "B45309"),
        LanguageInfo(code: "ak", name: "Akan", nativeName: "Akan", flag: "\u{1F1EC}\u{1F1ED}", colorHex: "EA580C"),

        // Central African / Cameroon
        LanguageInfo(code: "ln", name: "Lingala", nativeName: "Lingála", flag: "\u{1F1E8}\u{1F1E9}", colorHex: "2563EB"),
        LanguageInfo(code: "bas", name: "Bassa", nativeName: "\u{0181}\u{00E0}s\u{00E0}a", flag: "\u{1F1E8}\u{1F1F2}", colorHex: "DC2626"),
        LanguageInfo(code: "byv", name: "Medumba", nativeName: "M\u{0259}d\u{00F9}mb\u{00E0}", flag: "\u{1F1E8}\u{1F1F2}", colorHex: "16A34A"),
        LanguageInfo(code: "dua", name: "Douala", nativeName: "Duálá", flag: "\u{1F1E8}\u{1F1F2}", colorHex: "F59E0B"),
        LanguageInfo(code: "ewo", name: "Ewondo", nativeName: "Ewondo", flag: "\u{1F1E8}\u{1F1F2}", colorHex: "7C3AED"),
        LanguageInfo(code: "fan", name: "Fang", nativeName: "Fang", flag: "\u{1F1E8}\u{1F1F2}", colorHex: "0EA5E9"),

        // East & Southern African
        LanguageInfo(code: "sw", name: "Swahili", nativeName: "Kiswahili", flag: "\u{1F1F0}\u{1F1EA}", colorHex: "0D9488"),
        LanguageInfo(code: "zu", name: "Zulu", nativeName: "isiZulu", flag: "\u{1F1FF}\u{1F1E6}", colorHex: "1E40AF"),
        LanguageInfo(code: "xh", name: "Xhosa", nativeName: "isiXhosa", flag: "\u{1F1FF}\u{1F1E6}", colorHex: "0369A1"),
        LanguageInfo(code: "sn", name: "Shona", nativeName: "chiShona", flag: "\u{1F1FF}\u{1F1FC}", colorHex: "15803D"),
        LanguageInfo(code: "rw", name: "Kinyarwanda", nativeName: "Ikinyarwanda", flag: "\u{1F1F7}\u{1F1FC}", colorHex: "0EA5E9"),
        LanguageInfo(code: "rn", name: "Kirundi", nativeName: "Ikirundi", flag: "\u{1F1E7}\u{1F1EE}", colorHex: "DC2626"),
        LanguageInfo(code: "lg", name: "Luganda", nativeName: "Luganda", flag: "\u{1F1FA}\u{1F1EC}", colorHex: "CA8A04"),
        LanguageInfo(code: "so", name: "Somali", nativeName: "Soomaali", flag: "\u{1F1F8}\u{1F1F4}", colorHex: "2563EB"),
        LanguageInfo(code: "mg", name: "Malagasy", nativeName: "Malagasy", flag: "\u{1F1F2}\u{1F1EC}", colorHex: "16A34A"),
    ]

    // MARK: - Interface Languages (UI only)

    public static let interfaceLanguages: [LanguageInfo] = [
        LanguageInfo(code: "fr", name: "French", nativeName: "Français", flag: "\u{1F1EB}\u{1F1F7}", colorHex: "3B82F6"),
        LanguageInfo(code: "en", name: "English", nativeName: "English", flag: "\u{1F1EC}\u{1F1E7}", colorHex: "6366F1"),
        LanguageInfo(code: "es", name: "Spanish", nativeName: "Español", flag: "\u{1F1EA}\u{1F1F8}", colorHex: "EF4444"),
        LanguageInfo(code: "ar", name: "Arabic", nativeName: "\u{0627}\u{0644}\u{0639}\u{0631}\u{0628}\u{064A}\u{0629}", flag: "\u{1F1F8}\u{1F1E6}", colorHex: "15803D"),
    ]

    // MARK: - Lookup

    public static func info(for code: String) -> LanguageInfo? {
        allLanguages.first(where: { $0.code == code })
    }
}

File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/LanguageData.kt
package me.meeshy.sdk.model

/** Static metadata for a language — port of LanguageInfo (LanguageData.swift). */
data class LanguageInfo(
    val code: String,
    val name: String,
    val nativeName: String,
    val flag: String,
    val colorHex: String,
)

/** Static language tables — port of LanguageData (LanguageData.swift). */
object LanguageData {

    /** All translation-target languages. */
    val allLanguages: List<LanguageInfo> = listOf(
        // Romance
        LanguageInfo("fr", "French", "Francais", "🇫🇷", "3B82F6"),
        LanguageInfo("es", "Spanish", "Espanol", "🇪🇸", "EF4444"),
        LanguageInfo("it", "Italian", "Italiano", "🇮🇹", "22C55E"),
        LanguageInfo("pt", "Portuguese", "Portugues", "🇧🇷", "16A34A"),
        LanguageInfo("ro", "Romanian", "Romana", "🇷🇴", "2563EB"),
        LanguageInfo("ca", "Catalan", "Catala", "🏴", "EAB308"),
        // Germanic
        LanguageInfo("en", "English", "English", "🇬🇧", "6366F1"),
        LanguageInfo("de", "German", "Deutsch", "🇩🇪", "F59E0B"),
        LanguageInfo("nl", "Dutch", "Nederlands", "🇳🇱", "F97316"),
        LanguageInfo("sv", "Swedish", "Svenska", "🇸🇪", "0EA5E9"),
        LanguageInfo("da", "Danish", "Dansk", "🇩🇰", "DC2626"),
        LanguageInfo("no", "Norwegian", "Norsk", "🇳🇴", "1D4ED8"),
        LanguageInfo("af", "Afrikaans", "Afrikaans", "🇿🇦", "059669"),
        // Slavic
        LanguageInfo("ru", "Russian", "Russkij", "🇷🇺", "DC2626"),
        LanguageInfo("uk", "Ukrainian", "Ukrainska", "🇺🇦", "FBBF24"),
        LanguageInfo("pl", "Polish", "Polski", "🇵🇱", "E11D48"),
        LanguageInfo("cs", "Czech", "Cestina", "🇨🇿", "1E40AF"),
        LanguageInfo("sk", "Slovak", "Slovencina", "🇸🇰", "1E3A8A"),
        LanguageInfo("bg", "Bulgarian", "Balgarski", "🇧🇬", "15803D"),
        LanguageInfo("hr", "Croatian", "Hrvatski", "🇭🇷", "B91C1C"),
        LanguageInfo("sr", "Serbian", "Srpski", "🇷🇸", "9F1239"),
        LanguageInfo("sl", "Slovenian", "Slovenscina", "🇸🇮", "0369A1"),
        // Baltic
        LanguageInfo("lt", "Lithuanian", "Lietuviu", "🇱🇹", "CA8A04"),
        LanguageInfo("lv", "Latvian", "Latviesu", "🇱🇻", "7C2D12"),
        // Finno-Ugric
        LanguageInfo("fi", "Finnish", "Suomi", "🇫🇮", "2563EB"),
        LanguageInfo("hu", "Hungarian", "Magyar", "🇭🇺", "B45309"),
        LanguageInfo("et", "Estonian", "Eesti", "🇪🇪", "0284C7"),
        // Hellenic
        LanguageInfo("el", "Greek", "Ellinika", "🇬🇷", "1D4ED8"),
        // Turkic
        LanguageInfo("tr", "Turkish", "Turkce", "🇹🇷", "EF4444"),
        LanguageInfo("az", "Azerbaijani", "Azerbaycan", "🇦🇿", "0891B2"),
        LanguageInfo("kk", "Kazakh", "Kazaksa", "🇰🇿", "0EA5E9"),
        LanguageInfo("uz", "Uzbek", "O'zbek", "🇺🇿", "0D9488"),
        // Semitic
        LanguageInfo("ar", "Arabic", "al-arabiyyah", "🇸🇦", "15803D"),
        LanguageInfo("he", "Hebrew", "ivrit", "🇮🇱", "1D4ED8"),
        LanguageInfo("am", "Amharic", "amarinya", "🇪🇹", "16A34A"),
        // Indo-Aryan
        LanguageInfo("hi", "Hindi", "hindi", "🇮🇳", "F97316"),
        LanguageInfo("bn", "Bengali", "bangla", "🇧🇩", "059669"),
        LanguageInfo("ur", "Urdu", "urdu", "🇵🇰", "16A34A"),
        LanguageInfo("ne", "Nepali", "nepali", "🇳🇵", "DC2626"),
        LanguageInfo("ta", "Tamil", "tamil", "🇮🇳", "D97706"),
        // Iranian
        LanguageInfo("fa", "Persian", "farsi", "🇮🇷", "059669"),
        // Caucasian
        LanguageInfo("ka", "Georgian", "kartuli", "🇬🇪", "B91C1C"),
        LanguageInfo("hy", "Armenian", "hayeren", "🇦🇲", "EA580C"),
        // East Asian
        LanguageInfo("zh", "Chinese", "zhongwen", "🇨🇳", "DC2626"),
        LanguageInfo("ja", "Japanese", "nihongo", "🇯🇵", "E11D48"),
        LanguageInfo("ko", "Korean", "hangugeo", "🇰🇷", "1E40AF"),
        // Southeast Asian
        LanguageInfo("th", "Thai", "thai", "🇹🇭", "7C3AED"),
        LanguageInfo("vi", "Vietnamese", "Tieng Viet", "🇻🇳", "DC2626"),
        LanguageInfo("id", "Indonesian", "Bahasa Indonesia", "🇮🇩", "EF4444"),
        LanguageInfo("ms", "Malay", "Bahasa Melayu", "🇲🇾", "1D4ED8"),
        LanguageInfo("tl", "Filipino", "Filipino", "🇵🇭", "2563EB"),
        LanguageInfo("my", "Burmese", "myanmar", "🇲🇲", "CA8A04"),
        LanguageInfo("km", "Khmer", "khmer", "🇰🇭", "1E3A8A"),
        LanguageInfo("lo", "Lao", "lao", "🇱🇦", "B91C1C"),
        // West African
        LanguageInfo("yo", "Yoruba", "Yoruba", "🇳🇬", "16A34A"),
        LanguageInfo("ig", "Igbo", "Igbo", "🇳🇬", "15803D"),
        LanguageInfo("ha", "Hausa", "Hausa", "🇳🇬", "059669"),
        LanguageInfo("wo", "Wolof", "Wolof", "🇸🇳", "0D9488"),
        LanguageInfo("bm", "Bambara", "Bamanankan", "🇲🇱", "CA8A04"),
        LanguageInfo("ff", "Fulah", "Fulfulde", "🇸🇳", "0891B2"),
        LanguageInfo("tw", "Twi", "Twi", "🇬🇭", "D97706"),
        LanguageInfo("ee", "Ewe", "Ewegbe", "🇬🇭", "B45309"),
        LanguageInfo("ak", "Akan", "Akan", "🇬🇭", "EA580C"),
        // Central African / Cameroon
        LanguageInfo("ln", "Lingala", "Lingala", "🇨🇩", "2563EB"),
        LanguageInfo("bas", "Bassa", "Basaa", "🇨🇲", "DC2626"),
        LanguageInfo("byv", "Medumba", "Medumba", "🇨🇲", "16A34A"),
        LanguageInfo("dua", "Douala", "Duala", "🇨🇲", "F59E0B"),
        LanguageInfo("ewo", "Ewondo", "Ewondo", "🇨🇲", "7C3AED"),
        LanguageInfo("fan", "Fang", "Fang", "🇨🇲", "0EA5E9"),
        // East & Southern African
        LanguageInfo("sw", "Swahili", "Kiswahili", "🇰🇪", "0D9488"),
        LanguageInfo("zu", "Zulu", "isiZulu", "🇿🇦", "1E40AF"),
        LanguageInfo("xh", "Xhosa", "isiXhosa", "🇿🇦", "0369A1"),
        LanguageInfo("sn", "Shona", "chiShona", "🇿🇼", "15803D"),
        LanguageInfo("rw", "Kinyarwanda", "Ikinyarwanda", "🇷🇼", "0EA5E9"),
        LanguageInfo("rn", "Kirundi", "Ikirundi", "🇧🇮", "DC2626"),
        LanguageInfo("lg", "Luganda", "Luganda", "🇺🇬", "CA8A04"),
        LanguageInfo("so", "Somali", "Soomaali", "🇸🇴", "2563EB"),
        LanguageInfo("mg", "Malagasy", "Malagasy", "🇲🇬", "16A34A"),
    )

    /**
     * The codes shipped as interface (UI-chrome) languages, in display order. Derived
     * view over [allLanguages] so every language's metadata lives in exactly one place.
     */
    val interfaceLanguageCodes: List<String> = listOf("fr", "en", "es", "ar")

    /** UI-only interface languages, derived from [allLanguages] (no hand-copied drift). */
    val interfaceLanguages: List<LanguageInfo> =
        interfaceLanguageCodes.mapNotNull { code -> allLanguages.firstOrNull { it.code == code } }

    /**
     * The most frequently picked content languages, in the order they should surface at
     * the top of long pickers (onboarding / regional-language picker).
     */
    val commonLanguageCodes: List<String> = listOf(
        "fr", "en", "es", "de", "it", "pt", "ar", "zh", "ja", "ko",
        "ru", "tr", "nl", "pl", "sv", "hi", "th", "vi", "uk", "ro",
    )

    /**
     * [allLanguages] reordered so [commonLanguageCodes] lead in their declared order and
     * every other language follows in canonical order. A permutation — single base,
     * single ordering, nothing dropped or duplicated.
     */
    val allLanguagesCommonFirst: List<LanguageInfo> = run {
        val commonSet = commonLanguageCodes.toSet()
        val common = commonLanguageCodes.mapNotNull { code -> allLanguages.firstOrNull { it.code == code } }
        val rest = allLanguages.filterNot { it.code in commonSet }
        common + rest
    }

    /** Canonical-code aliases so legacy/BCP-47 spellings resolve (e.g. "fil" → "tl"). */
    private val aliases: Map<String, String> = mapOf("fil" to "tl")

    /**
     * Metadata for [code], or `null` when it is blank/absent or matches no known language.
     * Trimmed and case-insensitive; legacy aliases (e.g. "fil" → "tl") are resolved so
     * callers never need a local `lowercase()` / re-implemented matcher.
     */
    fun info(code: String?): LanguageInfo? {
        val cleaned = code?.trim()?.lowercase()?.takeIf { it.isNotEmpty() } ?: return null
        allLanguages.firstOrNull { it.code == cleaned }?.let { return it }
        val canonical = aliases[cleaned] ?: return null
        return allLanguages.firstOrNull { it.code == canonical }
    }
}
